packages/react-reconciler/src/ReactFiberNewContext.old.js

```diff
@@ -311,6 +329,101 @@ export function propagateContextChange<T>(
   }
 }
 
+export function lazilyPropagateParentContextChanges(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+) {
+  propagateParentContextChanges(current, workInProgress, renderLanes, false);
+}
+
+export function propagateParentContextChangesToDeferredTree(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+) {
+  propagateParentContextChanges(current, workInProgress, renderLanes, true);
+}
+
+function propagateParentContextChanges(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+  // TODO: This argument is currently unused. I think there's a way to optimize
+  // for the many providers case, where if the first propagation finds a match,
+  // the second one can avoid scanning down that same path. The trouble is that
+  // there could be a nested bailout below that.
+  forcePropagateEntireTree: boolean,
+) {
+  if (!enableLazyContextPropagation) {
+    return false;
+  }
+
+  let parent = workInProgress;
+  while (parent !== null && (parent.flags & DidPropagateContext) === NoFlags) {
+    if (parent.tag === ContextProvider) {
+      const currentParent = parent.alternate;
+      if (currentParent !== null) {
+        const oldProps = currentParent.memoizedProps;
+        if (oldProps !== null) {
+          const providerType: ReactProviderType<any> = parent.type;
+          const context: ReactContext<any> = providerType._context;
+
+          const newProps = parent.pendingProps;
+          const newValue = newProps.value;
+
+          const oldValue = oldProps.value;
+
+          const changedBits = calculateChangedBits(context, newValue, oldValue);
+          if (changedBits !== 0) {
+            // The context value changed. Search for matching consumers and
+            // schedule them to update.
+            propagateContextChange(
+              workInProgress,
+              context,
+              changedBits,
+              renderLanes,
+            );
+          }
+        }
+      }
+    }
+    parent = parent.return;
```

Comment:
This comment is not related to the others where I'm articulating a different approach, it's just an independent finding

There is an edge case where on any given bailout multiple stacked providers of the same context will be visited before you bail out on root or `DidPropagateContext`. In these cases changed bits are recalculated even for providers fibers that aren't top of stack. it won't technically break anything b/c the context value itself will always read consistently but it could lead to propagations that do not actually need to happen